server/src/Core.hs

```diff
@@ -74,70 +90,116 @@ type ServerState = SubscriptionTree UUID WS.Connection
-  tlogrecords <- newTBQueueIO queueCapacity
-  pure (Core tvalue tqueue tupdates tclients tlogrecords config metrics)
+  -- load the persistent data from disk
+  let filePath = configDataFile config
+  eitherValue <- readData filePath metrics
+  for eitherValue $ \initialValue -> do
+    -- create synchronization channels
+    tvalue <- newTVarIO initialValue
+    tdirty <- newTVarIO False
+    tqueue <- newTBQueueIO queueCapacity
+    tupdates <- newTBQueueIO queueCapacity
+    tclients <- newMVar newServerState
+    tlogrecords <- newTBQueueIO queueCapacity
+    pure (Core tvalue tdirty tqueue tupdates tclients tlogrecords config metrics)
 
 -- Tell the put handler loop, the update handler and the logger loop to quit.
 postQuit :: Core -> IO ()
 postQuit core = do
   atomically $ do
-    writeTBQueue (coreQueue core) Nothing
+    writeTBQueue (coreQueue core) Stop
     writeTBQueue (coreUpdates core) Nothing
     writeTBQueue (coreLogRecords core) Nothing
 
-enqueueOp :: Op -> Core -> IO EnqueueResult
-enqueueOp op core = atomically $ do
+-- | Try to enqueue a command. It succeeds if the queue is not full, otherwise,
+-- nothing is changed.
+tryEnqueueCommand :: Command -> Core -> IO EnqueueResult
+tryEnqueueCommand op core = atomically $ do
   isFull <- isFullTBQueue (coreQueue core)
-  unless isFull $ writeTBQueue (coreQueue core) (Just op)
+  unless isFull $ writeTBQueue (coreQueue core) op
   pure $ if isFull then Dropped else Enqueued
 
+-- | Enqueue a command. Blocks if the queue is full.
+enqueueCommand :: Command -> Core -> IO ()
+enqueueCommand cmd core = atomically $ writeTBQueue (coreQueue core) cmd
+
 getCurrentValue :: Core -> Path -> IO (Maybe Value)
 getCurrentValue core path =
   fmap (Store.lookup path) $ atomically $ readTVar $ coreCurrentValue core
 
 withCoreMetrics :: MonadIO m => Core -> (Metrics.IcepeakMetrics -> IO ()) -> m ()
 withCoreMetrics core act = liftIO $ forM_ (coreMetrics core) act
 
--- Execute an operation.
-handleOp :: Op -> Value -> Value
-handleOp op value = case op of
+-- Execute a modification.
+applyModification :: Modification -> Value -> Value
+applyModification op value = case op of
   Delete path -> Store.delete path value
   Put path newValue -> Store.insert path newValue value
 
--- Drain the queue of operations and apply them. Once applied, publish the
--- new value as the current one.
-processOps :: Core -> IO Value
-processOps core = atomically (readTVar (coreCurrentValue core)) >>= go
+-- | Drain the command queue and execute them. Changes are published to all
+-- subscribers. This function returns when executing the 'Stop' command from the
+-- queue.
+runCommandLoop :: Core -> IO Value
+runCommandLoop core = atomically (readTVar (coreCurrentValue core)) >>= go
   where
     go val = do
-      maybeOp <- atomically $ readTBQueue (coreQueue core)
-      case maybeOp of
-        Just op -> do
-          let newValue = handleOp op val
+      command <- atomically $ readTBQueue (coreQueue core)
```

Comment:
`atomically $ readTBQueue (coreQueue core)` shouldn't be _naked_ here. Is there no function already doing this? It seems like it should.

Reply:
I like it this way. It makes it immediately visible that this is an STM transaction (that could possibly be extended with more actions).